Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -1176,15 +1174,22 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
                                     localCache: documentationCache,
                                     moduleName: moduleName
                                 )
-                            } else if relationship.kind == .overloadOf {
-                                // An 'overloadOf' relationship points from symbol -> group
-                                overloadGroups[relationship.target, default: []].append(relationship.source)
                             }
                         }
-
-                        addOverloadGroupReferences(overloadGroups: overloadGroups)
                     }
-                    
+
+                    let overloadGroups: [String: Set<String>] =
+                    unifiedSymbolGraph.relationshipsByLanguage.values.joined().filter({
+                        $0.kind == .overloadOf
+                    }).map({
+                        (key: $0.target, value: $0.source)
+                    }).reduce([:], { acc, kv in
+                        var acc = acc
+                        acc[kv.key, default: []].insert(kv.value)
+                        return acc
+                    })
```

Comment:
I do admit i was particularly Haskell-brained when i wrote these PRs, but i'm not sure i want to use this version that creates multiple intermediate variable bindings just to break up the code.

@d-ronnqvist What do you think? I could have sworn there would be a better Dictionary initializer i could use, but this `reduce` was the closest thing i could find.

Reply:
I believe that it's possible to use [init(_:uniquingKeysWith:)][1] to combine the values, like this

[1]: https://developer.apple.com/documentation/swift/dictionary/init(_:uniquingkeyswith:)

```swift
let overloadGroups = [String: Set<String>](
    unifiedSymbolGraph.relationshipsByLanguage.values.joined().filter({
        $0.kind == .overloadOf
    }).map({
        (key: $0.target, value: $0.source)
    }),
    uniquingKeysWith: { $0.union($1) }
)
```

Using `reduce(into:_:)` is also fine but `reduce(_:_:)` with an unlabeled first parameter will end up making copies of the Set for each iteration of the loop. This accidentally makes an _O(n)_ looking implementation _O(n²)_.

This was actually the motivation for adding `reduce(into:_:)` in [SE-0171][2]:

[2]: https://github.com/apple/swift-evolution/blob/main/proposals/0171-reduce-with-inout.md

> The current version of `reduce needs to make copies of the result type for each element in the sequence. The proposed version can eliminate the need to make copies (when the `inout` is optimized away).

Regarding the sequence portion; I suspect that only a very small number of relationships will be `.overloadOf`. As such, I suspect that it's faster to filter the allocations first and join them after. It's also possible to combine `filter` and `map` into one iteration using `compactMap`:

```swift
let overloadRelationships: [(String, String)] = unifiedSymbolGraph.relationshipsByLanguage.values.flatMap {
    $0.compactMap {
        guard $0.kind == .overloadOf else { return nil }
        return (key: $0.target, value: $0.source)
    }
}
let overloadGroups = [String: Set<String>](overloadRelationships, uniquingKeysWith: { $0.union($1) })
```

Compared to this, using `reduce(into:_:)` has a slight readability benefit because it means that the value doesn't need to be mapped to a key-value-pair.

```swift
let overloadGroups: [String: Set<String>] = unifiedSymbolGraph.relationshipsByLanguage
    .values
    .flatMap {
        $0.filter { $0.kind == .overloadOf }
    }.reduce(into: [:]) { acc, rel in
        acc[rel.target, default: []].insert(rel.source)
    }
```

Reply:
You could also move the filter and loop inside the `reduce(into:_:)` closure:

```swift
let overloadGroups: [String: Set<String>] = unifiedSymbolGraph.relationshipsByLanguage
    .values
    .reduce(into: [:]) { acc, relationships in
        for rel in relationships where rel.kind == .overloadOf {
            acc[rel.target, default: []].insert(rel.source)
        }
    }
```

Reply:
Aha! Thanks for the note about `reduce(into:_:)`. I was a bit concerned about rebinding the accumulator like that; i'm glad there's an alternative that can also save an allocation.